### ISSUE TITLE
Using abstract ids for expressions/aliases/references (instead of URLs)

### ISSUE THREAD

**Issue by DavidBruant** [MyWebIntelligence/MyWebIntelligence#134]
# Current situation

[A single SQL table](https://github.com/MyWebIntelligence/MyWebIntelligence/blob/69a1948554c53f69a0d998b74f07b379f3d0f5c9/postgresDB/createTables.sql#L3-L12) is used to store the entire sub-web we save for a given territoire.

An Expression has a numerical id, but is more closely identified by its `uri`. `aliases` list all alias URIs (redirects for now, meta-canonical and other cases eventually). `references` links all **uris** being linked to.

## Recreating the graph

QueryResults are stored as a list of URLs returned by the oracle. This is on purpose. It's a good thing for historization and else to save things as they are returned by the oracles with no modification whatsoever.

Currently, these first URLs are used to get the first set of Expressions by comparisons with the `uri` and `aliases` fields. The `uri` can benefit from being indexed. The `aliases` array might be indexable for O(log(n)) searches but may require non-trivial work (apparently, B-Trees don't work, because they're a whole-value thing. There exists other type of indexes that may be better suited for arrays).

After the first expressions are acquired, a set of the URLs they `reference` is built and the Expression table is queried again by `uri`+`aliases`, etc. until natural exhaustion.

## Downsides

The index on `uri` most likely does it job well. Not having a proper index on `aliases` is likely the cause of several-second graph extractions even when the number of expressions is in the 10000s

More work into finding the right index type could work.

# Possible alternate plan

* Keep QueryResults as they are
* Make a table binding urls and an id either to an expression, or an alias
* Make a table for references (source, target)

```sql
CREATE TABLE IF NOT EXISTS expressions (
    id           SERIAL PRIMARY KEY,
    main_html    text,
    main_text    text,
    title        varchar(2000),
    meta_description   text
);

CREATE TABLE IF NOT EXISTS resources (
    id          SERIAL PRIMARY KEY,
    url         varchar(2000) UNIQUE NOT NULL,
    alias_of    integer UNIQUE REFERENCES resources (id), -- nullable if it doesn't alias anything
    expression  integer UNIQUE REFERENCES expressions (id) -- NULL initially and when uri is an alias
)
-- Per http://www.postgresql.org/docs/9.4/static/indexes-unique.html Postgresql already create all the necessary indices via the UNIQUE constraints

CREATE TABLE IF NOT EXISTS references (
    source      integer REFERENCES resources (id),
    target      integer REFERENCES resources (id)
)

-- to find graph edges quickly
CREATE INDEX references_from ON references (source);
```

* When query results come back
    * (create a QueryResults entry with the URLs)
    * create as many "blank" `resources` as there are URLs
    * (create the getExpressionTasks)
* When an expression is made from the network
    * update the `resource` if the URL was an alias
    * create a new `resource` if the URL was an alias
    * create an `expression`, update the corresponding `resource.expression` column with the id
    * create as many "blank" `resource` as there are new URLs in the expression links. Create corresponding `references` entries.
* When getting the graph for a given territoire
    * get all urls in corresponding QueryResults
    * get all `resources` for these URLs (n*log(m) with `resources.url` UNIQUE index)
        * If resource is an expression, find what it `references` (n*log(m) thanks to references_from index) (creates a new set of ids)
        * If the resource is an alias, follow this alias (add to the new set of ids)
        * If the resource is nothing, do nothing 'cause there is nothing to do
    * iterate with the new id set
    * On the server side, keep track of all URLs and ids, get rid of aliases eventually

For now, let's do the graph with multiple back-and-forth with the server. Should work well enough for now especially with all the new indices we get for free.

**Comment by DavidBruant** [comment]
Previously, absence of a URL in DB was also meaning "it's been visited, but 404 or content-type unusable". Now, resources are created as soon as the URL is once met.
To build a graph and avoiding 404s and PDFs, http_status needs to be recorded.

**Comment by DavidBruant** [comment]
Closed by #136.
